### PR TITLE
Add HCC Installation Detector

### PR DESCRIPTION
**Pull request by vsytch**
This change adds the HccInstallationDetector class. During initialisation it looks for a valid hcc installation and then adds the corresponding compilation and link arguments.

By default it looks in the directory from which hcc was invoked and in the directory in which clang(that's pointed by the invoked hcc) is. To manually specify a hcc installation add `--hcc-path=/path/to/hcc` to your arguments. 

We now always link to `libmcwamp.a` and `libhc_am.so`. To not link to them add `-fno-hcc-bootstrap` to your arguments.

### CODE REVIEW

File: include/clang/Driver/Options.td [hunk 1]
```diff
@@ -498,6 +498,12 @@ def cxxamp_cpu_mode : Flag<["-"], "cpu">, Flags<[DriverOption]>,Group<cxxamp_Gro
   HelpText<"C++AMP only.  This option allows the compiler to emit CPU kernel-specific artifacts"> ;
 def amdgpu_target_EQ : Joined<["-", "--"], "amdgpu-target=">, Flags<[DriverOption]>,
   HelpText<"Specify AMDGPU ISA version (ex: gfx803). Could be specified only once.">;
+def hcc_path_EQ : Joined<["--"], "hcc-path=">, Group<i_Group>,
+  HelpText<"HCC installation path">;
+def fhcc_bootstrap : Flag<["-"], "fhcc-bootstrap">,
+  HelpText<"Link to HCC runtime">;
```
Comment (scchan) [inline include/clang/Driver/Options.td]: The option description is inconsistent with the option name.
Either we change the description text for `fhcc-bootstrap` to `Use in hcc bootstrap build` or we change this option to some `fno-hcc-lib-link`

File: include/clang/Driver/Options.td [hunk 2]
```diff
@@ -498,6 +498,12 @@ def cxxamp_cpu_mode : Flag<["-"], "cpu">, Flags<[DriverOption]>,Group<cxxamp_Gro
...
+def fno_hcc_bootstrap : Flag<["-"], "fno-hcc-bootstrap">,
+  HelpText<"Don't link to HCC runtime">;
```
Comment (scchan) [inline include/clang/Driver/Options.td]: It seems like the HelpText is wrong and should be swapped with the previous option?

We already need a switch for `fno-hcc-bootstrap` ?

File: lib/Driver/ToolChains/Gnu.h [hunk 3]
```diff
@@ -19,6 +20,8 @@
 namespace clang {
 namespace driver {
 
+class HCCInstallationDetector;
```
Comment (scchan) [inline lib/Driver/ToolChains/Gnu.h]: This class is defined in Hcc.h so this seems to be redundant.

File: lib/Driver/ToolChains/Gnu.h [hunk 4]
```diff
@@ -62,7 +65,8 @@ class LLVM_LIBRARY_VISIBILITY Assembler : public GnuTool {
 };
 
 class LLVM_LIBRARY_VISIBILITY Linker : public GnuTool {
-  mutable std::unique_ptr<Linker> HCLinker;
+  mutable std::unique_ptr<Tool> HCLinker;
+  friend class HCC::CXXAMPLink;
```
Comment (scchan) [inline lib/Driver/ToolChains/Gnu.h]: I believe the reason for this friend class is allow `CXXAMPLink` to call the `ConstructLinkerJob`?  If that's the case, is it really necessary to call `gnutools::Linker::ConstructLinkerJob` from `CXXAMPLink::ConstructJob`?
In other words, could `CXXAMPLink` just return the additional arg back to `gnutools::Linker`  for it to call its own `ConstructLinkerJob`?

**Review by scchan — CHANGES_REQUESTED** [review]
A lot of the PSTL tests are failing, could you take a look?  Here's an example:


In file included from /home/scchan/code/hcc/tests/Unit/ParallelSTL/unique_copy_carray.cpp:6:
/home/scchan/code/build.hcc.vsytch-hcc_installation/compiler/bin/../../include/experimental/algorithm:27:10: fatal error: 'impl/algorithm_impl.inl' file not found
#include "impl/algorithm_impl.inl"
         ^~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.

**Comment by whchung** [comment]
@vsytch per discussion please update this PR after reflecting changes suggested by @scchan , and also please submit PRs against `hcc` for required changes in CMake or other components inside `hcc`.

**Review by scchan — APPROVED** [review]
LGTM

**Pull request merged**